### PR TITLE
Visibility field

### PR DESCRIPTION
**Pull request by phadej**
This is small step towards adding `visible` field for libraries.

This is **must** for the next release, so keep bikeshedding to the minimum. Concrete change proposals are most appreciated. I hope to merge this soon.

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by m-renaud** [comment]
High level comment: having just a boolean for visibility seems pretty limited. Most systems that I've worked with support a variety of visibility attributes ("public", "private", "test-only", "package", "custom", etc.) so it seems like limiting to just a boolean may prohibit some uses of this field in the future. The current change could remain as simple as it is today by changing the field name to `visibility` and supporting two values: `public` and `private` (corresponding to `visible: True` and `visible: False`).

**Comment by phadej** [comment]
I agree that `visibility: public` or `visibility: private` would make sense.

Other visibilities are out-of-scope for 3.0.  Would `test-only` mean that only test-suites in other packages would be able to depend on it. What's the use-case, is it worth complexity? Too much to think about. Not now.

**Comment by m-renaud** [comment]
> Other visibilities are out-of-scope for 3.0.

Perfectly reasonable.

> Would test-only mean that only test-suites in other packages would be able to depend on it

This wasn't the best example tbh, so lets forget about it for now. The "enumerate the list of packages that can see this" is the more interesting use case, which comes up a lot in a large code base. For example, when you only want to allow _some_ teams to depend on your code (because its not general availability yet) its incredibly useful to be able to lock down who can use the code. This would look something like:
```
package-group my-feature-internal-packages:
  my-feature-internal-lib
  my-feature-internal-tests

package-group my-feature-beta-users:
  package-a
  package-b

visibility: 
  my-feature-internal-packages
  my-feature-beta-users
```

**Comment by phadej** [comment]
ping @fgaz @hvr

**Comment by fgaz** [comment]
+1 for visibility: public|private

Actually I already coded this part locally, with the visibility ADT but without the tests. I wish I published it earlier :-/

**Comment by hvr** [comment]
I'm +1 for a non-boolean `visibility: (public | private )` (even if it remained a binary choice)

**Comment by phadej** [comment]
Changed to `visibility: public | private`

I made it so that in main library stanza `visibility` field is not even parsed (and therefore never printed either).

File: Cabal/Distribution/Types/LibraryVisibility.hs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    pretty LibraryVisibilityPublic  = Disp.text "public"
+    pretty LibraryVisibilityPrivate = Disp.text "private"
+
+instance Parsec LibraryVisibility where
+  parsec = do
+    name <- P.munch1 isAlpha
+    case name of
+      "public"  -> return LibraryVisibilityPublic
+      "private" -> return LibraryVisibilityPrivate
+      _         -> fail $ "Unknown visiblity: " ++ name
```
Comment (phadej) [inline Cabal/Distribution/Types/LibraryVisibility.hs]: typo

File: Cabal/Distribution/Types/Library.hs [hunk 2]
```diff
@@ -9,22 +9,24 @@ module Distribution.Types.Library (
...
 import qualified Distribution.Types.BuildInfo.Lens as L
 
 data Library = Library
     { libName           :: Maybe UnqualComponentName
     , exposedModules    :: [ModuleName]
     , reexportedModules :: [ModuleReexport]
-    , signatures        :: [ModuleName]   -- ^ What sigs need implementations?
-    , libExposed        :: Bool           -- ^ Is the lib to be exposed by default?
+    , signatures        :: [ModuleName]       -- ^ What sigs need implementations?
+    , libExposed        :: Bool               -- ^ Is the lib to be exposed by default? (i.e. registered into package db).
```
Comment (phadej) [inline Cabal/Distribution/Types/Library.hs]: (i.e. whether its modules available in GHCi for example)

File: Cabal/Distribution/Types/Library.hs [hunk 3]
```diff
@@ -36,31 +38,41 @@ instance Binary Library
...
+    , libVisibility     = mempty
+    , libBuildInfo      = mempty
+    }
+
+-- | This instance is not good.
+--
+-- We need it for 'PackageDescription.Configuration.addBuildableCondition'.
+-- More correct method would be some kind of "create empty clone".
+--
+-- More concretely, 'addBuildableCondition' will make `libVisibility = False`
```
Comment (fgaz) [inline Cabal/Distribution/Types/Library.hs]: s/False/LibraryVisibilityPrivate/

File: Cabal/Distribution/Types/Library.hs [hunk 4]
```diff
@@ -9,22 +9,24 @@ module Distribution.Types.Library (
...
 
 data Library = Library
     { libName           :: Maybe UnqualComponentName
     , exposedModules    :: [ModuleName]
     , reexportedModules :: [ModuleReexport]
-    , signatures        :: [ModuleName]   -- ^ What sigs need implementations?
-    , libExposed        :: Bool           -- ^ Is the lib to be exposed by default?
+    , signatures        :: [ModuleName]       -- ^ What sigs need implementations?
+    , libExposed        :: Bool               -- ^ Is the lib to be exposed by default? (i.e. whether its modules available in GHCi for example)
+    , libVisibility     :: LibraryVisibility  -- ^ Whether this multilib can be dependent from outside.
```
Comment (fgaz) [inline Cabal/Distribution/Types/Library.hs]: I think the comment should be "[...]can be depended on from outside"

Comment (hvr) [reply Cabal/Distribution/Types/Library.hs]: We could make it even clearer what is meant by "outside" by saying something like "... can be depended on from other packages"

**Comment by phadej** [comment]
I'm merging this as is, as finally CI is green. Will push `[skip ci]` textual fixes later

**Pull request merged**